### PR TITLE
[GLCC]Part-3: Support jit.save and jit.load for pylayer op

### PR DESCRIPTION
**Pull request by MarioLulab**
### PR types

New features

### PR changes

Others

#### Original requirements

It is expected that the PyLayer mechanism in the dynamic graph of the Paddle can interoperate with the @to_staitc of the Paddle's dynamic to static, supporting the custom layer of the PyLayer in the model to be perceived by the @to_static and correctly generating the static graph Program, and supporting the training of the dynamic to static and the exporting for inference. 

So our main objective is to **Enable PyLayer to Support @to_static**. In this PR, we support `pylayer` op can be exported by `jit.save` and loaded by `jit.load` for subsequent inference. 

#### Related issues

【Feature Request】 PyLayer 功能支持动转静 @to_static 🚀: https://github.com/PaddlePaddle/Paddle/issues/54120

#### Related PR

[GLCC]Part-1: Add pylayer op to Support @to_static: https://github.com/PaddlePaddle/Paddle/pull/56108

[GLCC]Part-2: @to_static support for PyLayer: https://github.com/PaddlePaddle/Paddle/pull/56531

#### Project Records

https://github.com/MarioLulab/GLCC2023-Paddle-Record

#### Main works

1. Enable `core.prune` and `core.prune_backward` to handle `BLOCKS` attribute of operators by refining the `prune.cc`.
2. Modify `normalize_program` function to remove backward block id of `pylayer` op.
3. Add some testcases in `test_prune.py`, `test_program_prune_backward.py` and `test_inference_model_io.py`, etc.
4. Support non-Variable input arguments and keyword-arguments for dy2st `PyLayer`.

### CODE REVIEW

File: test/dygraph_to_static/test_pylayer.py [hunk 1]
```diff
@@ -117,6 +214,48 @@ def forward(self, data):
...
+    def __init__(self, in_size, out_size):
+        super().__init__()
+        self.linear1 = paddle.nn.Linear(in_size, out_size)
+        self.linear2 = paddle.nn.Linear(in_size, out_size)
+
+    @paddle.jit.to_static
+    def forward(self, x1, x2):
+        y1 = self.linear1(x1)
+        y2 = self.linear1(x2)
+        out = cus_tanh_2.apply(y1, func1=paddle.tanh)
```
Comment (Aurelius84) [inline test/dygraph_to_static/test_pylayer.py:240]:
```suggestion
        out = cus_tanh_2.apply(y1, paddle.tanh)
```
部分单测可以去掉形参指示，这样可以丰富单测样例

File: test/legacy_test/test_prune.py [hunk 2]
```diff
@@ -160,77 +211,83 @@ def _mock_guard(mock):
     base.Executor._prune_program = original
 
 
+def net1():
```
Comment (Aurelius84) [inline test/legacy_test/test_prune.py:214]:
```suggestion
def create_net1():
```

File: test/legacy_test/test_prune.py [hunk 3]
```diff
@@ -160,77 +211,83 @@ def _mock_guard(mock):
...
+    loss2 = paddle.nn.functional.cross_entropy(
+        input=y, label=label, reduction='none', use_softmax=False
+    )
+    loss2 = paddle.mean(x=loss2)
+    loss1.persistable = True
+    loss2.persistable = True
+    return x, y, label, loss1, loss2, w_param_attrs
+
+
+def net2():
```
Comment (Aurelius84) [inline test/legacy_test/test_prune.py:241]:
```suggestion
def create_net2():
```

**Review by Aurelius84 — APPROVED** [review]
LGTM，一些代码规范上小的 Comment 可以单独提PR fix

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

File: test/dygraph_to_static/test_pylayer.py (hunk 1 above)
Comment (MarioLulab) [reply test/dygraph_to_static/test_pylayer.py:240]: 好的，将在下一个 pr 进行修改

File: test/legacy_test/test_prune.py (hunk 2 above)
Comment (MarioLulab) [reply test/legacy_test/test_prune.py:214]: 好的，将在下一个 pr 进行修改

File: test/legacy_test/test_prune.py (hunk 3 above)
Comment (MarioLulab) [reply test/legacy_test/test_prune.py:241]: 好的，将在下一个 pr 进行修改
